src/Functions/array/arrayDotProduct.cpp

```diff
@@ -49,22 +210,114 @@ class ArrayDotProductImpl
+    {
+        Type sum = 0;
+
+        void accumulate(Type x, Type y)
+        {
+            sum += x * y;
+        }
+
+        void combine(const State<Type> & other_state)
+        {
+            sum += other_state.sum;
+        }
+
+        Type finalize()
+        {
+            return sum;
+        }
+    };
+
+#if USE_MULTITARGET_CODE
+    template <typename ResultType>
+    AVX512_FUNCTION_SPECIFIC_ATTRIBUTE static void accumulateCombine(
+        const ResultType * __restrict data_x,
+        const ResultType * __restrict data_y,
+        size_t i_max,
+        size_t & i,
+        State<ResultType> & state)
+    {
+        __m512 sums;
+        if constexpr (std::is_same_v<ResultType, Float32>)
+            sums = _mm512_setzero_ps();
+        else
+            sums = _mm512_setzero_pd();
+
+        const size_t n = (std::is_same_v<ResultType, Float32>) ? 16 : 8;
+
+        for (; i + n < i_max; i += n)
+        {
+            if constexpr (std::is_same_v<ResultType, Float32>)
+            {
+                __m512 x = _mm512_loadu_ps(data_x + i);
+                __m512 y = _mm512_loadu_ps(data_y + i);
+                sums = _mm512_fmadd_ps(x, y, sums);
+            }
+            else
+            {
+                __m512 x = _mm512_loadu_pd(data_x + i);
+                __m512 y = _mm512_loadu_pd(data_y + i);
+                sums = _mm512_fmadd_pd(x, y, sums);
+            }
+        }
+
+        if constexpr (std::is_same_v<ResultType, Float32>)
+            state.sum = _mm512_reduce_add_ps(sums);
+        else
+            state.sum = _mm512_reduce_add_pd(sums);
+    }
+#endif
+
+    template <typename ResultType, typename LeftType, typename RightType>
+    static NO_SANITIZE_UNDEFINED ResultType apply(
+        const LeftType * left,
+        const RightType * right,
         size_t size)
     {
-        ResultType result = 0;
-        for (size_t i = 0; i < size; ++i)
-            result += static_cast<ResultType>(left[i]) * static_cast<ResultType>(right[i]);
-        return result;
+        State<ResultType> state;
+        size_t i = 0;
+
+        /// SIMD optimization: process multiple elements in both input arrays at once.
+        /// To avoid combinatorial explosion of SIMD kernels, focus on
+        /// - the two most common input/output types (Float32 x Float32) --> Float32 and (Float64 x Float64) --> Float64 instead of 10 x
+        ///   10 input types x 8 output types,
+        /// - the most powerful SIMD instruction set (AVX-512F).
+#if USE_MULTITARGET_CODE
+        if constexpr ((std::is_same_v<ResultType, Float32> || std::is_same_v<ResultType, Float64>)
+                        && std::is_same_v<ResultType, LeftType> && std::is_same_v<LeftType, RightType>)
+        {
+            if (isArchSupported(TargetArch::AVX512F))
+                accumulateCombine(left, right, size, i, state);
+        }
+#else
+        /// Process chunks in vectorized manner
+        static constexpr size_t VEC_SIZE = 4;
```

Comment:
I’m not familiar with vectorized execution. Just out of curiosity, why 4 is the best number of the loop size?

Reply:
In general: Dividing the loop into "sub-loops" helps the compiler to autovectorize better (see e.g. [here](https://llvm.org/docs/Vectorizers.html) for how that works in general). The thing is that - in theory - the compiler could/should chop up a loop by itself, without forcing the programmer to do it manually like here, but for some reason that did not happen for the original code:

```cpp
for (size_t i = 0; i < size; ++i)
    result += static_cast<ResultType>(left[i]) * static_cast<ResultType>(right[i]);
```

About `4`: The choice of the constant is sort of black magic. By default, ClickHouse is compiled with SSE 4.2 instructions as minimum SIMD instruction set (see [here](https://github.com/ClickHouse/ClickHouse/blob/master/cmake/cpu_features.cmake#L89)), which has 128-bit wide SIMD registers. So for input/output types Float32 / Float64 one would expect optimal constant values of 4 / 2. However, based on runtime measurements, the concrete choice (as long as it is a power of 2) does not make any difference in terms of performance ...

So, the only reason I chose `4` was that it was also chosen in similar code in a different function implementation (https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/array/arrayDistance.cpp#L580).

I'll add some more comments.

Reply:
Yeah, that might be!